x-pack/plugins/lens/public/xy_visualization/axis_settings_popover.tsx

```diff
@@ -0,0 +1,244 @@
+import { EuiIconAxisBottom } from '../assets/axis_bottom';
+// @ts-ignore
+import { EuiIconAxisLeft } from '../assets/axis_left';
+// @ts-ignore
+import { EuiIconAxisRight } from '../assets/axis_right';
+
+type AxesSettingsConfigKeys = keyof AxesSettingsConfig;
+export interface AxisSettingsPopoverProps {
+  /**
+   * Determines the popover title
+   */
+  popoverTitle: string;
+  /**
+   * Determines the axis
+   */
+  axis: AxesSettingsConfigKeys;
+  /**
+   * Contains the chart layers
+   */
+  layers?: LayerConfig[];
+  /**
+   * Determines the axis title
+   */
+  axisTitle: string | undefined;
+  /**
+   * Callback to axis title change
+   */
+  updateTitleState: (value: string) => void;
+  /**
+   * Determines the frame
+   */
+  frame: FramePublicAPI;
+  /**
+   * Determines if the popover is Disabled
+   */
+  isDisabled?: boolean;
+  /**
+   * Determines if the ticklabels of the axis are visible
+   */
+  areTickLabelsVisible: boolean;
+  /**
+   * Toggles the axis tickLabels visibility
+   */
+  toggleTickLabelsVisibility: (axis: AxesSettingsConfigKeys) => void;
+  /**
+   * Determines if the gridlines of the axis are visible
+   */
+  areGridlinesVisible: boolean;
+  /**
+   * Toggles the gridlines visibility
+   */
+  toggleGridlinesVisibility: (axis: AxesSettingsConfigKeys) => void;
+  /**
+   * Determines if the title visibility switch is on and the input text is disabled
+   */
+  isAxisTitleVisible: boolean;
+  /**
+   * Toggles the axis title visibility
+   */
+  toggleAxisTitleVisibility: (axis: AxesSettingsConfigKeys, checked: boolean) => void;
+}
+const popoverConfig = (
+  axis: AxesSettingsConfigKeys
+): { icon: IconType; groupPosition: ToolbarButtonProps['groupPosition'] } => {
+  switch (axis) {
+    case 'yLeft':
+      return {
+        icon: EuiIconAxisLeft,
+        groupPosition: 'left',
+      };
+    case 'yRight':
+      return {
+        icon: EuiIconAxisRight,
+        groupPosition: 'right',
+      };
+    case 'x':
+    default:
+      return {
+        icon: EuiIconAxisBottom,
+        groupPosition: 'center',
+      };
+  }
+};
+export const AxisSettingsPopover: React.FunctionComponent<AxisSettingsPopoverProps> = ({
+  layers,
+  popoverTitle,
+  axis,
+  axisTitle,
+  frame,
+  updateTitleState,
+  toggleTickLabelsVisibility,
+  toggleGridlinesVisibility,
+  isDisabled,
+  areTickLabelsVisible,
+  areGridlinesVisible,
+  isAxisTitleVisible,
+  toggleAxisTitleVisibility,
+}) => {
+  const [popoversOpenState, setPopoversOpenState] = useState(false);
+  const [title, setTitle] = useState<string | undefined>(axisTitle);
+
+  const getAxisTitle = useCallback(() => {
+    const defaultTitle = axisTitle;
+
+    const initialLayer = layers?.[0];
+    if (!initialLayer || !initialLayer.accessors.length) {
+      return defaultTitle;
+    }
+    const datasource = frame.datasourceLayers[initialLayer.layerId];
+    if (!datasource) {
+      return defaultTitle;
+    }
+    switch (axis) {
+      case 'yLeft':
```

Comment:
This logic assumes that there is always a "left" axis but does not assume there is always a "right" axis. It's possible to define a single axis on the right, so this looks wrong to me.

Reply:
Let's sync about it first when you are available

Reply:
You are right, will do a small refactor here

Reply:
So, I did a small refactor here to identify if there is a left or right axis as indeed there is the possibility to have a single right axis. But there are problems when the user starts 'playing' with the axis side and changes the settings to `auto`. In that case, I don't know which axis is active and which label to display. (and which popover to enable too)
Is there any way to know which legend is displayed in each axis? Because the computation becomes too complex and I don't know which is the best solution here.

cc @flash1293

Reply:
You can't always know which metric will end up on which side as it can depend on the formatter you don't know about in the configuration. Also, you can't always say whether an axis will be active or not.

If there are two metrics using auto, there is a chance to have a left and a right axis, but only if there are different formatters.

I think the best we can do is to use the current logic for the cases we can know of:
* Use existing axis assignment logic from `x-pack/plugins/lens/public/xy_visualization/axes_configuration.ts`, always assuming `isFormatterCompatible` will return false (because this is the worst case)
* This will tell you whether to enable a popover or not (it might be wrong in some cases, enabling an axis even if it's not necessary right now - we can't work around that at the moment - but it will never disable a popover if it's needed)
* For the logic in `getAxisTitle` - in the cases we can know for sure which dimension will end up first on the axis we can show that, otherwise we just show an empty input with "Overwrite axis title" placeholder. The logic to check that is the following:
  * Go through all accessors in the order like in `x-pack/plugins/lens/public/xy_visualization/axes_configuration.ts` (iterate over layers, in there iterate over accessors)
  * If you encounter an accessor with an explicit "left" axis config, this will become the title for the left axis. If you encounter another explicit "left" axis config after this one, just ignore.
  * If you encounter an accessor with an explicit "right" axis config, this will become the title for the left axis. If you encounter another explicit "right" axis config after this one, just ignore.
  * If you encounter an accessor with an "auto" axis config or no axis config, there are two possibilities:
    * There is no title set yet for the left axis by another accessor, in this case the current accessor will become the title for the left axis
    * Otherwise you can't know, and both left and right axis title are unknown (and should be shown as empty string)

Once we expose the current table to the visualization and data source, we can solve this better and rely on the currently rendered state.

Reply:
Just a thought - do we actually need to show the current axis title in the overwrite menu? Why not always show this:
<img width="312" alt="Screenshot 2020-09-09 at 18 30 48" src="https://user-images.githubusercontent.com/1508364/92626749-a3952e80-f2ca-11ea-90f3-c6ceb665a158.png">

This would simplify the logic a lot (saving the whole second part and basically reducing `getAxisTitle` to `() => ''`).

@cchaos Do you think that would be OK? It wouldn't change anything about the behavior, it's just about not prefilling the axis title input in the config popover with the currently active value.

Reply:
First of all, Joe thank you a lot about your solution. Secondly, I agree, the simpler the better. As there will be cases that we won't be able to cover, maybe it will be better to leave it blank for now.

Reply:
Totally fine with that for now

Reply:
I have changed the logic on fetching the axis labels with a different way and it seems to work on almost all the cases. If we are ok with it we can keep it, otherwise, let's make it simple as Joe proposed. cc @cchaos

Reply:
So we discussed it with Caroline and we will avoid this complexity for now. We will go with the placeholders